proxy/proxy.go

```diff
@@ -0,0 +1,705 @@
+package proxy
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io/ioutil"
+	"log"
+	"math/rand"
+	"net"
+	"net/http"
+	"os"
+	"os/signal"
+	"sync"
+	"syscall"
+	"time"
+
+	G "gopkg.in/gilmour-libs/gilmour-e-go.v4"
+	"gopkg.in/gilmour-libs/gilmour-e-go.v4/backends"
+)
+
+// LogError prints error on console, returns string format of the error and raises a panic
+func LogError(err error) string {
+	panic(err)
+	log.Println(err.Error())
+	return err.Error()
+}
+
+// Status is a integer to hold status of node of the node (unavailable = 0, ok = 1 and dirty = 2)
+type Status int
+
+// Status Constant
+const (
+	Unavailable Status = 0 + iota
+	Ok
+	Dirty
+)
+
+var status = [...]string{"unavailable", "ok", "dirty"}
+
+// String returns the English name of the status ("unavailable", "ok", ...).
+func (s Status) String() string { return status[s] }
+
+// NodeID is a string to hold node's id
+type NodeID string
+
+// GilmourTopic is a string to hold topic for signal or request
+type GilmourTopic string
+
+// implements NodeMapOperations
+type nodeMap struct {
+	sync.Mutex
+	regNodes map[NodeID]*Node
+}
+
+func initNodeMap() (nm *nodeMap) {
+	nm = new(nodeMap)
+	nm.regNodes = make(map[NodeID]*Node)
+	return
+}
+
+var (
+	nMap = initNodeMap()
+)
+
+// NodeMapOperations is a interface to enable operation on nodeMap
+type NodeMapOperations interface {
+	Put(NodeID, *Node) error
+	Del(NodeID) error
+	Get(NodeID) (*Node, error)
+}
+
+// GetNodeMap returns nodeMap
+func GetNodeMap() nodeMap {
+	return *nMap
+}
+
+// Put adds node in nodeMap
+func (n *nodeMap) Put(uid NodeID, node *Node) (err error) {
+	n.Mutex.Lock()
```

Comment:
Good practice is to have `defer n.Mutex.Unlock()` right after you acquire a lock.